x/programs/rust/examples/automated-market-maker/src/lib.rs

```diff
@@ -1,261 +1,249 @@
+    let lp_token = external_liquidity_token(&context);
+    let lp_balance = token::balance_of(&lp_token, context.actor());
+    remove_liquidity(context, lp_balance)
+}
 
-        let program_id = simulator
-            .run_step(&Step::create_program(PROGRAM_PATH))
-            .unwrap()
-            .id;
+/// Returns the token reserves in the pool
+fn reserves(token_x: &ExternalCallContext, token_y: &ExternalCallContext) -> (u64, u64) {
+    let balance_x = token::allowance(
+        token_x,
+        *token_x.program().account(),
+        *token_x.program().account(),
+    );
+    let balance_y = token::allowance(
+        token_y,
+        *token_y.program().account(),
+        *token_y.program().account(),
+    );
 
-        let test_context = TestContext::from(program_id);
+    (balance_x, balance_y)
+}
 
-        let resp = simulator
-            .run_step(&Step {
-                endpoint: Endpoint::Execute,
-                method: "add_liquidity".to_string(),
-                max_units: u64::MAX,
-                params: vec![test_context.clone().into(), 1000u64.into(), 1000u64.into()],
-            })
+/// Returns the tokens in the pool
+fn token_programs(context: &Context<StateKeys>) -> (Program, Program) {
+    (
+        context
+            .get(StateKeys::TokenX)
             .unwrap()
-            .result
-            .response::<u64>()
-            .unwrap();
-
-        assert_eq!(resp, 1000);
-
-        simulator
-            .run_step(&Step {
-                endpoint: Endpoint::Execute,
-                method: "swap".to_string(),
-                max_units: u64::MAX,
-                params: vec![test_context.clone().into(), 10u64.into(), true.into()],
-            })
-            .unwrap();
-
-        let (amount_x, amount_y) = simulator
-            .run_step(&Step {
-                endpoint: Endpoint::Execute,
-                method: "remove_liquidity".to_string(),
-                max_units: u64::MAX,
-                params: vec![test_context.into(), 1000.into()],
-            })
+            .expect("token x not initialized"),
+        context
+            .get(StateKeys::TokenY)
             .unwrap()
-            .result
-            .response::<(u64, u64)>()
-            .unwrap();
+            .expect("token y not initialized"),
+    )
+}
+/// Returns the external call contexts for the tokens in the pool
+fn external_token_contracts(
+    context: &Context<StateKeys>,
+) -> (ExternalCallContext, ExternalCallContext) {
+    let (token_x, token_y) = token_programs(context);
+
+    (
+        ExternalCallContext::new(token_x, MAX_GAS, 0),
+        ExternalCallContext::new(token_y, MAX_GAS, 0),
+    )
+}
+
+/// Returns the external call context for the liquidity token
+fn external_liquidity_token(context: &Context<StateKeys>) -> ExternalCallContext {
+    ExternalCallContext::new(
+        context.get(StateKeys::LiquidityToken).unwrap().unwrap(),
+        MAX_GAS,
+        0,
+    )
+}
```

Comment:
nit: this would panic with an unwrap if the program is not initialized

```suggestion
fn external_liquidity_token(context: &Context<StateKeys>) -> ExternalCallContext {
    let liquidity_token = context.get(StateKeys::LiquidityToken).unwrap().expect("liquidity token not initialized");
    ExternalCallContext::new(liquidity_token, MAX_GAS, 0)
}
```